### ISSUE TITLE
Add example to show how to compute correlation between 2 4D images

### ISSUE THREAD

**Issue by jyeatman** [nilearn/nilearn#2948]
Akin to nilearn.image.mean_img I think there should be nilearn.image.corr_img that would take in two 4D images and returns a 3D image of correlation coefficients. 

This would be particularly useful for (a) computing the reliability of each voxel's timeseries in an fMRI experiment with multiple repeats of the same stimulus sequence and (b) computing reliability for naturalistic viewing paradigms. For example determining which voxels show a reliable response to multiple repeats of movie viewing.

### Pseudocode for the new behavior, if applicable

```python

```

**Comment by GaelVaroquaux** [comment]
Hi @jyeatman ,

I think that you should be able to use math_img for such a functionality.

**Comment by jyeatman** [comment]
Thanks @GaelVaroquaux 
That was my first thought but I couldn't figure out how to wrap that into a 1 line np operation. Could you provide an example? Otherwise I'm happy to submit a PR with image.corr_img (though I don't know the nilearn code base well so might not adhere to the standards)

**Comment by bthirion** [comment]
Can't make it a one-liner either, but I may have missed something.

**Comment by jyeatman** [comment]
Is the idea that math_img is used for numpy one liners and that a separate routine would be written for anything that requires more than a single numpy command?

**Comment by GaelVaroquaux** [comment]
I agree that writing the numpy one-liner in the general case might be a
bit hard (maybe impossible, I'd like to pick up the challenge, but AAAI
deadline is tonight).

If you input only standardized images, it should be a question
of doing something like np.sum(img1 * img2, axis=-1).

With non-normalized images, it's more like
np.sum(img1 * img2, axis=-1) / np.sqrt(np.sum(img1 **2, axis=-1) * np.sum(img2 **2, axis=-1))

Forgive me, I haven't actually run those lines. Maybe I'm talking
nonsense.

**Comment by jyeatman** [comment]
I think that one liner would require de-meaning the images first. I couldn't get it close to one line myself

```
def my_dot(x, y):
    return np.sum(x * y, -1)
def ndcor(data1,data2):
    data1 = data1-np.expand_dims(np.mean(data1, axis=-1),axis=3)
    data2 = data2-np.expand_dims(np.mean(data2, axis=-1),axis=3)
    corr = my_dot(data1, data2) / (np.sqrt(my_dot(data1, data1)) * np.sqrt(my_dot(data2, data2)))
    return corr
```

**Comment by bthirion** [comment]
Indeed. 
Of course, we cannot create another new function for each operation one wants to perform on images, but  in that case it sounds acceptable to me: the functional call for a relatively basic and straightfoward operation would replace  several lines of code.

**Comment by jyeatman** [comment]
Plus I think this is a generally useful function that has many use cases as opposed to one that is specific to a given project

Sent from my iPhone
________________________________

**Comment by NicolasGensollen** [comment]
Sounds acceptable to me too.
@jyeatman do you want to work on a PR ?

**Comment by tsalo** [comment]
From what I understand from [Afyouni, Smith, & Nichols (2019)](https://doi.org/10.1016/j.neuroimage.2019.05.011), statistical inference on correlations of autocorrelated time series (e.g., fMRI data) is affected by effective degrees of freedom. Assuming I understand correctly, would it be worth it to include methods from that paper (implemented in [`asoroosh/xDF`](https://github.com/asoroosh/xDF)) for calculating accurate z-statistics? Or maybe just include a warning in the docstring that a simple Fisher's z-transform is insufficient for the resulting correlation coefficient map?

**Comment by bthirion** [comment]
+1 for a warning that correlation values does not map straightforwardly to p-values,a nd that we discourage drawing statistical conclusions from such correlations.

**Comment by jyeatman** [comment]
The method suggested by @tsalo sound super useful https://github.com/asoroosh/xDF

@tsalo did you want to try to implement (I think it’s beyond my Python skills)? Or should I submit a simple PR with the correlation calculation and the warning message?

**Comment by tsalo** [comment]
I started playing around with the code on a fork and it's pretty heavily built around ROI-to-ROI analyses. I think it's going to take me a while to figure out how to get the code working for seed-to-voxel or voxel-to-voxel correlations, so it might be best to start with the warning and tackle inference in a later PR.

**Comment by bthirion** [comment]
I think I would discourage rather than try and implement, this feature. Resting-state correlations are too  easily confounded by  type of spurious effects that they should never be converted to p-values. What matters instead is the comparison of such correlations across subjects, groups etc.
(I know that a fraction of the community loves to do this kind of things, but to me this is a dead end).

**Comment by tsalo** [comment]
I hadn't thought of it like that. I won't strongly advocate in favor of the correction, since I don't know enough about its use-cases or limitations. For the warning, then, I think it would be important to not just say that statistical inference on individual correlation maps is discouraged, but also that the appropriate way to analyze the results would be in a group analysis, for which one necessary intermediate step would be to convert to z-values (not test z-statistics, but just a `numpy.arctanh`).

@jyeatman does that work for you?

**Comment by jyeatman** [comment]
Yes that totally makes sense to me. I agree completely with @bthirion 's points. I think most use cases would use correlations as an effect size / reliability estimate and wouldn't use p-values anyway (unless computed across subjects as suggested by @bthirion ).

**Comment by htwangtw** [comment]
So I want to address this issue on two fronts. Long story short - I don't think nilearn should implement this very specific function. 

1. How to calculate voxel-to-voxel correlation - I would advise against implementing the specific feature for two reasons:

- The most conceptually straightforward workflow is using `NiftiMasker` to get voxel-level timeseries, and calculate the correlation with `ConnectivityMeasure`, given you have a big enough RAM. If we implement a one-liner for this, it's wrapping two pretty big objects under one, which I would discourage. This use case can be addressed by a tutorial. 
 
- To make this computation efficient without introducing new dependencies, we will need to implement some mathematical tricks that are very specific to memory efficiency. 

- We have been having conversations related to performance optimisation in the nilearn team, and this issue will be solved once we have a path forward for performance improvement. The current plan is to follow the direction of `scikit-learn`. There's an example plugin that they are working on, AFAIK: https://github.com/soda-inria/sklearn-numba-dpex 

2. Relationship with xDF - I would separate this issue from the xDF issue, as xDF can be potentially useful in other contexts that are not limited to voxel-to-voxel connectivity. I will move the discussion there. 

cc @bthirion @Remi-Gau @mtorabi59

**Comment by bthirion** [comment]
I agree with your approach.

**Comment by htwangtw** [comment]
Shall we conclude that we won't implement this and close the issue?

**Comment by Remi-Gau** [comment]
> Shall we conclude that we won't implement this and close the issue?

For now I changed the title of the issue to make it clear that we are now thinking of adding an example.

**Comment by bthirion** [comment]
IMHO it's OK to close.
